nydusify: introduce chunkdict generate subcommand

newthifans opened the pull request:
### Notice
Please take a look and give me some code suggestions in the commit, thanks!

### Brief introduction
Add the function of storing multiple nydus images into the database for generating chunkdict
Implement the command “nydusify chunkdict generate --sources”
by invoking the command “nydus-image chunkdict save --bootstrap” .

 ### **Basic usage**
nydusify chunkdict generate --sources localhost:5000/ubuntu:nydus_latest,localhost:5000/ubuntu:nydus_latest_v1

### Details
1. Write command “nydusify chunkdict generate” entries
2. Create a “chunkdict/generator” folder to implement the function
3. Store the image in the registry with slices and Store them in the database in turn
4. Parse nydus image and create a directory to hold corresponding bootstrap
5. Pull nydus image bootstrap and store  corresponding directory
6. Invoke  the command “nydus-image chunkdict save --bootstrap”

This PR is related to #1305

imeoer commented on contrib/nydusify/cmd/nydusify.go:
```diff
@@ -632,6 +633,68 @@ func main() {
+					Name:  "generate",
+					Usage: "Save chunk and blob information of Multi-image into the database (experimental)",
+					Flags: []cli.Flag{
+						&cli.StringFlag{
```
Does it work if use `StringSliceFlag`?

imeoer replied:
Otherwise, we need to remind the user to use comma-separated image names.

imeoer commented on contrib/nydusify/pkg/chunkdict/generator/bootstrap.go:
```diff
@@ -0,0 +1,34 @@
+	if err := builder.Save(build.SaveOption{
+		BootstrapPath: bootstrap.BootstrapPath,
+	}); err != nil {
+		return errors.Wrap(err, "invalid nydus bootstrap format")
```
Why say invalid format?

imeoer commented on contrib/nydusify/pkg/chunkdict/generator/generator.go:
```diff
@@ -0,0 +1,107 @@
+// "Chunkdict generate" command generate chunkdict by deduplicating multiple nydus images
+// by invoking "nydus-image chunkdict save" command to save chunk and blob information image into database.
+// and by invoking "nydus-image chunkdict generate" to perform a deduplication algorithm from information of database to chunkdict
+type Generator struct {
```
`Generator generates blabla...`

imeoer commented on contrib/nydusify/pkg/chunkdict/generator/generator.go:
```diff
@@ -0,0 +1,107 @@
+	sourcesParser []*parser.Parser
+}
+
+// New creates Generator instance, sourceSlice is multiple Nydus images reference list.
```
What's `sourceSlice`?

imeoer commented on contrib/nydusify/pkg/chunkdict/generator/generator.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+// "Chunkdict Generate" deposit multiple Nydus images into the database one by one.
+func (generator *Generator) Generate(ctx context.Context) error {
```
`Generate saves`

imeoer commented on contrib/nydusify/pkg/chunkdict/generator/generator.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+// "Save" function stores information of chunk and blob of a Nydus Image in the database
+func (generator *Generator) save(ctx context.Context, index int) error {
```
`save stores...`

imeoer commented on contrib/nydusify/pkg/chunkdict/generator/generator.go:
```diff
@@ -0,0 +1,107 @@
+		return errors.Wrap(err, "output image information")
+	}
+
+	bootstrap := Bootstrap{
```
I think it's unnecessary to put the chunkdict save logic in `Bootstrap`, it should be put into `generator.Generate` directly.

imeoer commented on contrib/nydusify/pkg/build/builder.go:
```diff
@@ -143,3 +147,17 @@ func (builder *Builder) Run(option BuilderOption) error {
 	return builder.run(args, option.PrefetchPatterns)
 }
+
+// Save calls `nydus-image save` to parse Nydus bootstrap
```
`nydus-image chunkdict generate` ?

imeoer replied:
`nydus-image chunkdict save`.

imeoer commented on contrib/nydusify/pkg/build/builder.go:
```diff
@@ -143,3 +147,17 @@ func (builder *Builder) Run(option BuilderOption) error {
 }
+
+// Save calls `nydus-image save` to parse Nydus bootstrap
+// and output debug information to specified JSON file.
```
What's the specified JSON file?

imeoer commented on contrib/nydusify/pkg/chunkdict/generator/generator.go:
```diff
@@ -0,0 +1,107 @@
+	return generator, nil
+}
+
+// "Generate save" deposit multiple Nydus images into the database one by one.
```
Generate saves multiple Nydus bootstraps into the database one by one.

imeoer commented on contrib/nydusify/pkg/chunkdict/generator/generator.go:
```diff
@@ -0,0 +1,107 @@
+
+// Generator generates chunkdict by deduplicating multiple nydus images
+// by invoking "nydus-image chunkdict save" command to save chunk and blob information image into database.
+// and by invoking "nydus-image chunkdict generate" to perform a deduplication algorithm from information of database to chunkdict
```
Please note the length limit for single-line comment and paragraph alignment.

imeoer commented on contrib/nydusify/pkg/chunkdict/generator/generator.go:
```diff
@@ -0,0 +1,107 @@
+
+// New creates Generator instance.
+func New(opt Opt) (*Generator, error) {
+	// TODO: support sources resolver
```
What's the sources resolver?

imeoer commented on contrib/nydusify/pkg/chunkdict/generator/generator.go:
```diff
@@ -0,0 +1,107 @@
+		return errors.Wrap(err, "output image information")
+	}
+
+	//Invoke "nydus-image save" command
```
Please add a space after `//`.

imeoer commented on contrib/nydusify/pkg/chunkdict/generator/generator.go:
```diff
@@ -0,0 +1,107 @@
+		return errors.Wrap(err, "invalid nydus bootstrap format")
+	}
+
+	logrus.Infof("Chunkdict successfully save Nydus image %s", generator.sourcesParser[index].Remote.Ref)
```
`Saved chunk information from image %s`

imeoer commented on contrib/nydusify/pkg/chunkdict/generator/output.go:
```diff
@@ -0,0 +1,66 @@
+			return errors.Wrap(err, "unpack Nydus bootstrap layer")
+		}
+	} else {
+		err := fmt.Errorf("this is not Nydus image")
```
return fmt.Errorf("The %s is not a Nydus image")

imeoer commented on contrib/nydusify/cmd/nydusify.go:
```diff
@@ -632,6 +633,68 @@ func main() {
 		},
+		{
+			Name:  "chunkdict",
+			Usage: "Deduplicate RAFS filesystem metadata",
```
`Deduplicate chunk for Nydus image (experimental)`

newthifans committed:
nydusify chunkdict generate --sources
Add the 'nydus-image chunkdict save' command
with the "--sources" followed by the nydus image of registry
(e.g.,'registry.com/busybox:nydus-v1,registry.com/busybox:nydus-v2')

Signed-off-by: Zhao Yuan <1627990440@qq.com>
